### PR TITLE
Thchang/stokes dropdown in histogram and statistics

### PR DESCRIPTION
**Pull request by TienHao**
This PR is for #1171. Supporting the stokes dropdown in the histogram and statistics widgets. 

Companion branches:
backend: [mark/628_stokes_for_stat_and_hist_widgets](https://github.com/CARTAvis/carta-backend/tree/mark/628_stokes_for_stat_and_hist_widgets) (https://github.com/CARTAvis/carta-backend/pull/790)
protobuf: [mark/stokes_for_stat_and_hist_calc](https://github.com/CARTAvis/carta-protobuf/tree/mark/stokes_for_stat_and_hist_calc)

https://user-images.githubusercontent.com/20379662/125691432-2d6c33cc-ff61-4ffc-809a-e07340d5d421.mov

### CODE REVIEW

**Comment by kswang1029** [comment]
@TienHao would you turn this into a formal PR? I did a quick test and it appears fine. Will have a more detailed test once an issue is fixed in the corresponding backend branch.

**Comment by kswang1029** [comment]
@acdo2002 could you test this from a user point of view too (especially the correctness)?

**Comment by acdo2002** [comment]
@kswang1029  Ok I am going to test it.

**Comment by acdo2002** [comment]
I found the current dev, with a given stoke, change the channel, the statistics widget and histogram widget will update the information/plot.

However in this frontend branch, 
with a given stoke, then change the channel, the histogram widget only update at the first round, then stick there (the histogram widget does not update the plot but the statistics changes when the channel changes).

https://user-images.githubusercontent.com/34430648/126305528-d8215c02-1def-47e8-9358-f86c8f836601.mp4

**Comment by markccchiang** [comment]
> I found the current dev, with a given stoke, change the channel, the statistics widget and histogram widget will update the information/plot.
> 
> However in this frontend branch,
> with a given stoke, then change the channel, the histogram widget only update at the first round, then stick there (the histogram widget does not update the plot but the statistics changes when the channel changes).
> 

I think this is the backend's problem. Have fixed it and updated the backend branch. @acdo2002 please check if the problem is solved. Thank you!

**Comment by acdo2002** [comment]
I confirmed that the previous issue I posted has been solved. Thanks @markccchiang points out the backend issue. Now the missing cached histogram message has been disappeared. 

I think Stokes divides into "Current/I/Q/U/V" is good design, the user can compare the histogram/statistics (given a region or different regions) between different channel/stokes at the same time.

File: src/components/Histogram/HistogramComponent.tsx [hunk 1]
```diff
@@ -66,10 +67,17 @@ export class HistogramComponent extends React.Component<WidgetProps> {
...
-            const data = frameMap.get(regionId);
-            if (data && data.histograms && data.histograms.length) {
-                return data.histograms[0];
+            const regionMap = frameMap.get(regionId);
+            if (!regionMap) {
+                return null;
+            }
+            const stokes = this.widgetStore.effectiveFrame.stokesInfo.findIndex(stokes => stokes === coordinate.slice(0, 1));
+            const regionHistogramData = regionMap.get(stokes === -1 ? this.widgetStore.effectiveFrame.requiredStokes : stokes);
+            // const regionHistogramData = regionMap.get(coordinate);
```
Comment (YuHsuan-Hwang) [inline src/components/Histogram/HistogramComponent.tsx:76]: remove comment

Comment (YuHsuan-Hwang) [inline src/components/Histogram/HistogramComponent.tsx]: remove comment

File: src/components/Histogram/HistogramToolbarComponent/HistogramToolbarComponent.tsx [hunk 2]
```diff
@@ -1,15 +1,52 @@
...
+    }
+
     public render() {
+        const widgetStore = this.props.widgetStore;
+
+        let enableStokesSelect = false;
+        let stokesClassName = "unlinked-to-selected";
+        const coordinateOptions = [{value: "z", label: "Current"}];
+
+        if (widgetStore.effectiveFrame && widgetStore.effectiveFrame.regionSet) {
```
Comment (YuHsuan-Hwang) [inline src/components/Histogram/HistogramToolbarComponent/HistogramToolbarComponent.tsx]: may simplify to `widgetStore.effectiveFrame?.regionSet`

File: src/components/Stats/StatsComponent.tsx [hunk 3]
```diff
@@ -199,6 +216,23 @@ export class StatsComponent extends React.Component<WidgetProps> {
     public render() {
         const appStore = AppStore.Instance;
 
+        const widgetStore = this.widgetStore;
+
+        let enableStokesSelect = false;
+        let stokesClassName = "unlinked-to-selected";
+        const coordinateOptions = [{value: "z", label: "Current"}];
+
+        if (widgetStore.effectiveFrame && widgetStore.effectiveFrame.regionSet) {
```
Comment (YuHsuan-Hwang) [inline src/components/Stats/StatsComponent.tsx]: may simplify to `widgetStore.effectiveFrame?.regionSet`

File: src/stores/AppStore.ts [hunk 4]
```diff
@@ -1453,11 +1457,12 @@ export class AppStore {
         // Apply pending channel histogram
         const key = `${tileStreamDetails.fileId}_${tileStreamDetails.stokes}_${tileStreamDetails.channel}`;
         const pendingHistogram = this.pendingChannelHistograms.get(key);
-        if (pendingHistogram && pendingHistogram.histograms && pendingHistogram.histograms.length) {
+        if (pendingHistogram && pendingHistogram.histograms) {
```
Comment (YuHsuan-Hwang) [inline src/stores/AppStore.ts]: may simplify to `pendingHistogram?.histograms`

**Review by YuHsuan-Hwang — COMMENTED** [review]
minor changes

**Review by kswang1029 — APPROVED** [review]

**Review by veggiesaurus — APPROVED** [review]

**Comment by kswang1029** [comment]
@TienHao The stokes dropdown menu in the statistics and the histogram widgets does not work properly. It stays as current no matter which other options are selected. Could you have a look?

**Comment by YuHsuan-Hwang** [comment]
the problem should be caused by https://github.com/CARTAvis/carta-frontend/pull/1595. I can fix this

**Pull request merged**